### PR TITLE
ISPN-11676 + ISPN-11642 Hot Rod client per-cache config and on-access cache creation + JCache Remote URI

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.redhat.com/browse/ISPN-11676
Hot Rod client per-cache configuration and on-access cache creation

https://issues.redhat.com/browse/ISPN-11642
JCache Remote: URI pointing to a hotrod-client.properties file should be used to supply configuration

### CODE REVIEW

**Comment by tristantarrant** [comment]
Failures are unrelated

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/configuration/ConfigurationChildBuilder.java [hunk 1]
```diff
@@ -222,6 +222,8 @@
     */
    TransactionConfigurationBuilder transaction();
 
+   RemoteCacheConfigurationBuilder remoteCache(String name);
```
Comment (karesti) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/configuration/ConfigurationChildBuilder.java:225]: Java doc

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/RemoteCacheManager.java [hunk 2]
```diff
@@ -427,52 +430,84 @@ private Properties loadFromStream(InputStream stream) {
...
+            // We may be able to create it. Don't use RemoteCacheAdmin for this, since it would end up calling this method again
+            Map<String, byte[]> params = new HashMap<>(2);
+            params.put(RemoteCacheManagerAdminImpl.CACHE_NAME, cacheName.getBytes(HotRodConstants.HOTROD_STRING_CHARSET));
+            if (cacheConfiguration != null && cacheConfiguration.templateName() != null) {
+               params.put(RemoteCacheManagerAdminImpl.CACHE_TEMPLATE, cacheConfiguration.templateName().getBytes(HotRodConstants.HOTROD_STRING_CHARSET));
+            } else if (cacheConfiguration != null && cacheConfiguration.configuration() != null) {
+               params.put(RemoteCacheManagerAdminImpl.CACHE_CONFIGURATION, new XMLStringConfiguration(cacheConfiguration.configuration()).toXMLString(cacheName).getBytes(HotRodConstants.HOTROD_STRING_CHARSET));
+            } else {
+               // We cannot create the cache
                return null;
```
Comment (karesti) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/RemoteCacheManager.java:457]: instead if returning null and exception or logging that would explain better the issue ?

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/logging/Log.java [hunk 3]
```diff
@@ -332,4 +332,7 @@
 
    @Message(value = "OAUTHBEARER mechanism selected without providing a token", id = 4093)
    CacheConfigurationException oauthBearerWithoutToken();
+
+   @Message(value = "Cannot specify both template name and configuration for '%s'", id = 4094)
+   CacheConfigurationException remoteCacheTemplateNameXorConfiguration(String name);
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply client/hotrod-client/src/main/java/org/infinispan/client/hotrod/logging/Log.java:337]: exclusive or: you can have either one or the other

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/RemoteCacheManager.java (hunk 2 above)
Comment (tristantarrant) [reply client/hotrod-client/src/main/java/org/infinispan/client/hotrod/RemoteCacheManager.java:457]: I really would like to throw an exception, but the previous version returned null and I didn't want to break backwards compat

File: documentation/src/main/asciidoc/topics/proc_configuring_hotrod_createonaccess.adoc [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+----
+import org.infinispan.client.hotrod.configuration.ConfigurationBuilder;
+import org.infinispan.client.hotrod.configuration.NearCacheMode;
+...
+
+// Specify the server-side template or configuration to use
+ConfigurationBuilder builder = new ConfigurationBuilder();
+builder
+  .remoteCache("my-cache")
+    .templateName("org.infinispan.DIST_SYNC")
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply documentation/src/main/asciidoc/topics/proc_configuring_hotrod_createonaccess.adoc:17]: Done

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/Util.java [hunk 5]
```diff
@@ -28,20 +29,19 @@ public int getFormatId() {
 
       @Override
       public byte[] getGlobalTransactionId() {
-         return new byte[] {1};
+         return new byte[]{1};
       }
 
       @Override
       public byte[] getBranchQualifier() {
-         return new byte[] {1};
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/Util.java:36]: I've used our IDEA code style and that's what it does

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/configuration/ConfigurationChildBuilder.java (hunk 1 above)
Comment (tristantarrant) [reply client/hotrod-client/src/main/java/org/infinispan/client/hotrod/configuration/ConfigurationChildBuilder.java:230]: Done

**Comment by tristantarrant** [comment]
@oraNod some docs changes here

**Pull request merged**